sbe-tool/src/main/java/uk/co/real_logic/sbe/xml/XmlSchemaParser.java

```diff
@@ -326,6 +326,18 @@ public static String getAttributeValue(final Node elementNode, final String attr
         return attrNode.getNodeValue();
     }
 
+    /**
+     * To be used with child elements of {@code <types>} elements. Returns the package attribute as
+     * defined on the parent {@code <types>} element
+     *
+     * @param elementNode the node inside the types element
+     * @return the package name, or null if not defined
+     */
+    public static String getTypesPackageAttribute(final Node elementNode)
+    {
+        return getAttributeValue(elementNode.getParentNode(), "package", null);
```

Comment:
This is not likely to be sufficient for nested composites or ref types as the parent will be different.

Reply:
True. I need to write a test case for this. So we speak about two potential cases as below?

```xml
...
<messageSchema package="composite.elements"
               id="3"
               semanticVersion="5.2"
               description="Unit Test"
               byteOrder="littleEndian">
    <types package="specific.package.name">
        <composite name="messageHeader" description="Message identifiers and length of message root">
            <type name="blockLength" primitiveType="uint16"/>
            <type name="templateId" primitiveType="uint16"/>
            <type name="schemaId" primitiveType="uint16"/>
            <type name="version" primitiveType="uint16"/>
        </composite>
      ...
       <!-- case number 1 -->
       <enum name="booleanEnum" encodingType="uint8" semanticType="Boolean">
            <validValue name="F">0</validValue>
            <validValue name="T">1</validValue>
        </enum>
        <composite name="futuresPrice">
            <type name="mantissa" primitiveType="int64" />
            <type name="exponent" primitiveType="int8" />
            <ref name="isSettlement" type="booleanEnum" offset="10" />
        </composite>
 
      <!-- case number 2 -->
      <composite name="outer">
            <enum name="enumOne" description="enum as uint8" encodingType="uint8" semanticType="int">
                <validValue name="Value1">1</validValue>
                <validValue name="Value10">10</validValue>
            </enum>
            <type name="zeroth" primitiveType="uint8"/>
            <set name="setOne" description="set as uint32" encodingType="uint32" semanticType="int">
                <choice name="Bit0">0</choice>
                <choice name="Bit16">16</choice>
                <choice name="Bit26">26</choice>
            </set>
            <composite name="inner">
                <type name="first" primitiveType="int64"/>
                <type name="second" primitiveType="int64"/>
            </composite>
        </composite>
    </types>
...
```

Reply:
Yes these cases and consider the likes of: https://github.com/real-logic/simple-binary-encoding/blob/master/sbe-samples/src/main/resources/example-schema.xml#L32-L34

BTW I've pushed a change to better find the ancestor `<types>` element.